Code/Framework/AzCore/AzCore/Settings/CommandLineParser.h

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#pragma once
+
+#include <AzCore/std/containers/fixed_vector.h>
+#include <AzCore/std/containers/span_fwd.h>
+#include <AzCore/std/functional.h>
+#include <AzCore/std/string/fixed_string.h>
+#include <AzCore/std/string/string_view.h>
+#include <AzCore/std/utility/expected.h>
+
+namespace AZ::Settings
+{
+    //! Stores the prefix used for option arguments
+    //! The cap is MaxOptionPrefixes
+    //! The prefixes can't be larger than 4 characters
+    inline static constexpr size_t MaxOptionPrefixes = 8;
+    inline static constexpr size_t MaxOptionPrefixSize = 4;
+    using CommandLineOptionPrefixString = AZStd::basic_fixed_string<char, MaxOptionPrefixSize, AZStd::char_traits<char>>;
+    using CommandLineOptionPrefixArray = AZStd::fixed_vector<CommandLineOptionPrefixString, MaxOptionPrefixes>;
+    using CommandLineOptionPrefixSpan = AZStd::span<CommandLineOptionPrefixString const>;
+}
+
+namespace AZ::Settings::Platform
+{
+    CommandLineOptionPrefixArray GetDefaultOptionPrefixes();
+}
+
+namespace AZ::Settings
+{
+    //! Struct which contains a string view for option name and the value of a command line argument
+    //! If the command line argument is a positional argument, then the m_option string_view is empty
+    struct CommandLineArgument
+    {
+        AZStd::string_view m_option;
+        AZStd::string_view m_value;
+    };
+
+    //! Settings structure which is used to how to parse CLI command line parameters
+    //! It is setup to invoke an callback with each argument option name and value
+    //! to allow parsing of the command line with allocation of any heap memory
+    //! This can be used to parse the command line in order to read in settings
+    //! which can configure memory allocators before they are available without the need
+    //! to allocate any heap memory as long as the user supplied callback itself is careful
+    //! and how it stores it's settings.
+    //! Suggestion: Using structures such as fixed_vector and fixed_string, allows arrays and string like structures
+    //! that uses stack memory to store the command line settings in a structure for further use which avoids
+    //! any heap allocations
```

Comment:
```suggestion
    //! Suggestion: Use structures such as fixed_vector and fixed_string to store the settings
    //! in the user callback to avoid heap allocations and rely entirely on stack memory instead.
```